msg-common/src/span.rs

```diff
@@ -0,0 +1,185 @@
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use derive_more::{Deref, DerefMut};
+
+/// A container with a [`tracing::Span`] attached.
+///
+/// If `T` is a `Future`, the span will be entered when polling the future, making this struct
+/// almost equivalent to [`tracing::Instrument`]. The main difference is that the return value will
+/// contain the span as well so it can be re-entered later.
+#[derive(Debug, Clone, Deref, DerefMut)]
+pub struct WithSpan<T> {
+    #[deref]
+    #[deref_mut]
+    pub inner: T,
+    pub span: tracing::Span,
+}
+
+impl<T> WithSpan<T> {
+    /// Create a spanned container using [`tracing::Span::current`] span.
+    #[inline]
+    pub fn current(inner: T) -> Self {
+        Self { inner, span: tracing::Span::current() }
+    }
+
+    /// Create a container with a no-op [`tracing::Span::none`] span, to be eventually replaced
+    /// with [`Self::with_span`]
+    #[inline]
+    pub fn new(inner: T) -> Self {
+        Self { inner, span: tracing::Span::none() }
+    }
+
+    /// Replace the current [`tracing::Span`] with the provided one.
+    #[inline]
+    pub fn with_span(mut self, span: tracing::Span) -> Self {
+        self.span = span;
+        self
+    }
+
+    /// Break the spanned container into a tuple containing the inner object and the span.
+    #[inline]
+    pub fn into_parts(self) -> (T, tracing::Span) {
+        (self.inner, self.span)
+    }
+
+    #[inline]
+    pub fn into_inner(self) -> T {
+        self.inner
+    }
+}
+
+impl<T: Future> Future for WithSpan<T> {
+    type Output = WithSpan<T::Output>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        unsafe {
+            // SAFETY: we never move `inner` while polling, and span is `Unpin`.
+            let this = self.get_unchecked_mut();
+            let inner = Pin::new_unchecked(&mut this.inner);
+            let span = &this.span;
+
+            let _g = span.enter();
+
+            if let Poll::Ready(val) = inner.poll(cx) {
+                return Poll::Ready(val.with_span(span));
+            }
+
+            Poll::Pending
+        }
+    }
+}
+
+/// A container with a [`tracing::span::EnteredSpan`] attached.
+#[derive(Debug, Deref, DerefMut)]
+pub struct WithEntered<T> {
+    #[deref]
+    #[deref_mut]
+    pub inner: T,
+    pub span: tracing::span::EnteredSpan,
+}
+
+/// Trait to convert [`WithSpan<T>`] containers into [`WithEntered<S>`] containers, by
+/// [`Enter::enter`]ing the span.
```

Comment:
```suggestion
/// Trait mainly used to convert [`WithSpan<T>`] containers into [`WithEntered<S>`] containers, by
/// [`Span::enter`]ing the span.
```

Reply:
TBH there's nothing linking this trait to a span, or indicating that the source type should have a span.

In that sense, this is more like an `Into` with an associated type.
Maybe worth changing it to instead something like:
```rust
trait EnterSpan {
  type WithEntered;
  
  fn span(&self) -> &Span;
  fn with_entered(&self, _: EnteredSpan) -> WithEntered;

  fn enter(&self) -> WithEntered {
    let span = self.span();
    EnterSpan::with_entered(&self, span.entered())
   }
}
```

Yes more ugly but makes the trait purpose clear, without relying on on documentation alone to enforce the contract